### PR TITLE
Handle twap in orderbook

### PR DESCRIPTION
**Pull request by anxolin**
image

This PR implements and test the logic in TWAP that instructs the watch-towers what is the next polling time for the case, the current part is already in the order book. 

This is very important for the scalability of watch towers, since otherwise they would need to check on every block.
This PR will allow a DCA that executes once a month, to be polled only once a month (two at most, since first polling returns the current part, and next polling for the same period will instruct to wait for approximately one month )

The test will cover various edge cases in the execution of a 10 parts TWAP order. Read the tests for more details. As a summary:
- Will verify that when is the time of part N, it instructs watch tower to wait until N+1 starts
- Will verify that if we are in the last part, then it will instructs watch tower to not poll again this order
- It also tests some edge-cases that should not happen, but formalises the assumption of the responsibility of this function. This function is supposed to be called only if the current part is yielding an order that is already in the orderboook, so it shouldn’t be possible that the TWAP is expired or that is hasn’t started. However, the implementation handles these cases, and the test also covers them

## Test
`yarn test`

### CODE REVIEW

File: src/composable/orderTypes/Twap.ts [hunk 1]
```diff
@@ -350,11 +350,37 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
     const { numberOfParts } = this.data
     const startTimestamp = await this.startTimestamp(params)
 
+    if (blockTimestamp < startTimestamp) {
+      return {
+        result: PollResultCode.UNEXPECTED_ERROR,
```
Comment (alfetopito) [inline src/composable/orderTypes/Twap.ts:355]: Shouldn't this return the error indicating to try again at `x` block instead of unnexpected error?

File: src/composable/orderTypes/Twap.ts [hunk 2]
```diff
@@ -350,11 +350,37 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
...
-    const currentPartNumber = Math.floor(blockTimestamp - startTimestamp / timeBetweenParts)
+    const currentPartNumber = Math.floor((blockTimestamp - startTimestamp) / timeBetweenParts)
+
+    // If current part is the last one
+    if (currentPartNumber === numberOfParts.toNumber() - 1) {
+      return {
+        result: PollResultCode.DONT_TRY_AGAIN,
+        reason: `Current active TWAP part (${
+          currentPartNumber + 1
+        }/${numberOfParts}) is already in the Order Book. This was the last TWAP part, nor more orders need to be placed`,
```
Comment (alfetopito) [inline src/composable/orderTypes/Twap.ts:378]:
```suggestion
        }/${numberOfParts}) is already in the Order Book. This was the last TWAP part, no more orders need to be placed`,
```

File: src/composable/orderTypes/Twap.ts (hunk 1 above)
Comment (anxolin) [reply src/composable/orderTypes/Twap.ts:355]: Good question, but in my opinion it shouldn't. That's the responsibility of the `pollValidate` which should have returned the error you say, so no-one should have called this function.

Additionally, this function should be called after getting the order data, which would also fail for the same reason. 

Therefore, this `UNEXPECTED_ERROR` is just signaling that something unexpected happened, so it can be matched in DEV time.

File: src/composable/orderTypes/Twap.ts (hunk 2 above)
Comment (anxolin) [reply src/composable/orderTypes/Twap.ts]: Fixed here and in the test. good catch!

**Review by alfetopito — APPROVED** [review]

File: src/composable/orderTypes/Twap.ts (hunk 1 above)
Comment (mfw78) [reply src/composable/orderTypes/Twap.ts:355]: Given that this is an execution path that shouldn't occur (and thus as mentioned, is actually an _unexpected error_, perhaps a little note in the code for maintainers along the lines of "This shouldn't happen as `pollValidate` should have caught before attempting a poll", or something similar.

File: src/composable/orderTypes/Twap.ts [hunk 3]
```diff
@@ -350,11 +350,37 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
...
+      return {
+        result: PollResultCode.UNEXPECTED_ERROR,
+        reason: `TWAP is expired. Expired at ${expireTime} (${formatEpoch(expireTime)})`,
+        error: undefined,
+      }
+    }
+
     // Get current part number
-    const currentPartNumber = Math.floor(blockTimestamp - startTimestamp / timeBetweenParts)
+    const currentPartNumber = Math.floor((blockTimestamp - startTimestamp) / timeBetweenParts)
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.ts:370]: Disregard my comments in the PR #168  about this math, I see it's fixed here :+1: :guitar:

File: src/composable/orderTypes/Twap.ts [hunk 4]
```diff
@@ -363,9 +389,11 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
...
       result: PollResultCode.TRY_AT_EPOCH,
       epoch: nextPartStartTime,
-      reason: `Current active TWAP part (${currentPartNumber}/${numberOfParts}) is already in the Order Book. TWAP part ${currentPartNumber} doesn't start until ${nextPartStartTime} (${formatEpoch(
-        nextPartStartTime
-      )})`,
+      reason: `Current active TWAP part (${
+        currentPartNumber + 1
+      }/${numberOfParts}) is already in the Order Book. TWAP part ${
+        currentPartNumber + 2
+      } doesn't start until ${nextPartStartTime} (${formatEpoch(nextPartStartTime)})`,
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.ts:396]: Given debugging / logs here, I'd suggest that we make notes that TWAPs are indexed by '0', but logged with indexes commencing at 1. Alternatively (and one that may be simpler from a debugging perspective) is to report the part as zero indexed, and explicitly state the total. Something like:

`Current active TWAP part 0[5] is already in the Order Book. TWAP part 1[5] doesn't start until` etc. This would have the debugging messages remain consistent with what developers can expect (and potentially prevent silly off-by-one issues).

I have no strong opinion as to which one to prefer, just that we should either establish a standard (not sure if we have a standard in place for this already?), or keep logging indexes the same as that used within code.

File: src/composable/orderTypes/Twap.spec.ts [hunk 5]
```diff
@@ -333,3 +334,189 @@ describe('Poll Validate', () => {
...
+  test(`Polling at the start of part 2/10`, async () => {
+    // GIVEN: Part 2 just started
+    const pollParams = getPollParams({
+      blockTimestamp: startTimestamp + timeBetweenParts,
+    })
+
+    // WHEN: We invoke handlePollFailedAlreadyPresent
+    const result = await twap.handlePollFailedAlreadyPresent(orderId, order, pollParams)
+
+    // THEN: It should instruct we should wait for part 2 to start
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.spec.ts:445]:
```suggestion
    // THEN: It should instruct we should wait for part 3 to start
```

File: src/composable/orderTypes/Twap.spec.ts [hunk 6]
```diff
@@ -333,3 +334,189 @@ describe('Poll Validate', () => {
...
+  test(`Polling at the last second of part 9/10`, async () => {
+    // GIVEN: Part 9 is about to end
+    const pollParams = getPollParams({
+      blockTimestamp: startTimestamp + 9 * timeBetweenParts - 1,
+    })
+
+    // WHEN: We invoke handlePollFailedAlreadyPresent
+    const result = await twap.handlePollFailedAlreadyPresent(orderId, order, pollParams)
+
+    // THEN: It should instruct we should wait for part 2 to start
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.spec.ts:463]:
```suggestion
    // THEN: It should instruct we should wait for part 10 to start
```

File: src/composable/orderTypes/Twap.spec.ts [hunk 7]
```diff
@@ -333,3 +334,189 @@ describe('Poll Validate', () => {
...
+    expect(result).toEqual({
+      result: PollResultCode.TRY_AT_EPOCH,
+      reason:
+        "Current active TWAP part (9/10) is already in the Order Book. TWAP part 10 doesn't start until 1700000900 (2023-11-14T22:28:20.000Z)",
+      epoch: 1700000900,
+    })
+  })
+
+  test(`Polling at the first second of part 10/10`, async () => {
+    // GIVEN: Part 10 is about to end
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.spec.ts:473]:
```suggestion
    // GIVEN: Part 10 has just started
```

File: src/composable/orderTypes/Twap.spec.ts [hunk 8]
```diff
@@ -333,3 +334,189 @@ describe('Poll Validate', () => {
...
+    const pollParams = getPollParams({
+      blockTimestamp: startTimestamp + 9 * timeBetweenParts,
+    })
+
+    // WHEN: We invoke handlePollFailedAlreadyPresent
+    const result = await twap.handlePollFailedAlreadyPresent(orderId, order, pollParams)
+
+    // THEN: It should instruct we should wait for part 2 to start
```
Comment (mfw78) [inline src/composable/orderTypes/Twap.spec.ts:481]:
```suggestion
    // THEN: It should instruct that this was the last TWAP part.
```

**Review by mfw78 — APPROVED** [review]
Minor typos and nits

File: src/composable/orderTypes/Twap.ts (hunk 4 above)
Comment (anxolin) [reply src/composable/orderTypes/Twap.ts:396]: Good point, i will try to add some comments. I mean, the reason is just a human readable message, i see your point on being 0-indexed, but the SDK is offering some easier way to use the TWAP, this is why  it lets the user use`TwapData` instead of making them work with the `TwapStruct`, so my thinking is that is easier to understand 1-index based twap orders: there's 10 orders, 1/10, 2/10  is simpler than there's 10 orders 0/9 which is the first one, 1/9 which is the second one. 

This is just a message to see in the logs, so I think is fine, I will document is not 0-indexed though.

Let me know if you feel strong about this one
